spec/unit/job_fetcher_spec.rb

```diff
@@ -136,6 +136,62 @@ module Backend
             end
           end
         end
+
+        context 'when `wait` is set to true' do
+          let(:enqueued_job) { [job_queue, subject.encode(BackgroundJob.new)] }
+
+          it 'calls a blocking redis command' do
+            expect(test_redis).to receive(:blpop)
+            subject.fetch
+          end
+
+          context 'and `max` is nil' do
+            it 'returns one job' do
+              expect(test_redis).to receive(:blpop).and_return(enqueued_job)
+              result = subject.fetch
+
+              expect(result.class).to eq Resque::Job
+            end
+          end
+
+          context 'and `max` is not nil' do
+            it 'returns one job' do
+              expect(test_redis).to receive(:blpop).and_return(enqueued_job)
+              result = subject.fetch(max: 5)
+
+              expect(result.size).to eq 1
+            end
+          end
+        end
+
+        context 'when `wait` is set to false' do
+          let(:enqueued_jobs) {
+            [subject.encode(BackgroundJob.new), subject.encode(BackgroundJob.new), subject.encode(BackgroundJob.new)]
+          }
+
+          it 'calls a non-blocking redis command' do
+            expect(test_redis).to receive(:lpop).at_least(:once)
+            subject.fetch(wait: false)
+          end
+
+          context 'and `max` is nil' do
+            it 'returns all available jobs' do
+              expect(test_redis).to receive(:lpop).and_return(enqueued_jobs)
+              result = subject.fetch(wait: false)
+
+              expect(result.size).to eq 3
```

Comment:
In fact it should return a single job when `max` is `nil`, waiting or not waiting.

So this shows the point that while describing the behavior is useful, because we stub database logic, at the end we receive what we have expected.

This is because `max` is set to `1` when not set. 

https://github.com/3scale/apisonator/blob/a3234d650cbf84644214ff936afdee79d67faa67/lib/3scale/backend/job_fetcher.rb#L70

The test exposes though something that I didn't know what to do with. That is 

https://github.com/3scale/apisonator/blob/a3234d650cbf84644214ff936afdee79d67faa67/lib/3scale/backend/job_fetcher.rb#L37

Presently if database returned more results, then `#fetch` returns them. I was wondering whether to install any special error reporting. But thought that such error is in general should be a rather extreme database misbehavior that I don't see how we would be supposed to recover from.

The caller would probably fail, report the error and we will have an indication what happened.

So we can either keep it or maybe add special reporting if database returned more entries than requested.

Reply:
btw this test is the same as
https://github.com/3scale/apisonator/pull/382/files#diff-dcde5659e5ac647cc27329d886893e94a01e744c54eca60ccaff43f09ea489deL51-L55

Reply:
> In fact it should return a single job when max is nil, waiting or not waiting.

Right, the test was wrong, fixed.

> So this shows the point that while describing the behavior is useful, because we stub database logic, at the end we receive what we have expected.

Unit tests are about testing a unit independent from it's context, so stubs are needed. The problem is this test was wrong.

> this test is the same as https://github.com/3scale/apisonator/pull/382/files#diff-dcde5659e5ac647cc27329d886893e94a01e744c54eca60ccaff43f09ea489deL51-L55

It's not the same, that one is testing the returned job contains a valid value for the queue and payload, and it tests this in a particular context: when the job contain invalid UTF-8 characters. My test is part of a series of tests about what the method does according to received parameters. For instance, fetch returns a `Job` in some cases, and an `Array` in other cases, and the array can have one or many elements. What the method returns is not intuitive, e.g. It returns an array of one element when `wait` is `true` and `max` not `nil`, but it returns a `Job` when `wait` is `false` and `max` is `nil`. So depending on `wait`, it returns a class or another. The code is already merged and the caller depends on that behavior so we must test it and have tests that will break if that behavior changes.

> The test exposes though something that I didn't know what to do with. That is

That code is already merged, so that's what tests must test right now.